src/server/qgsconfigcache.cpp

```diff
@@ -23,27 +22,73 @@
 
 #include <QFile>
 
-QgsConfigCache *QgsConfigCache::instance()
+QgsConfigCache *QgsConfigCache::sInstance = nullptr;
+
+
+void QgsConfigCache::initialize( QgsServerSettings *settings )
 {
-  static QgsConfigCache *sInstance = nullptr;
+  if ( sInstance )
+  {
+    QgsMessageLog::logMessage(
+      QStringLiteral( "Project's cache is already initialized" ),
+      QStringLiteral( "Server" ), Qgis::MessageLevel::Warning );
+    return;
+  }
 
+  QgsAbstractCacheStrategy *strategy;
+  if ( settings && settings->projectCacheStrategy() == QStringLiteral( "periodic" ) )
+  {
+    strategy = new QgsPeriodicCacheStrategy( settings->projectCacheCheckInterval() );
+    QgsMessageLog::logMessage(
+      QStringLiteral( "Initializing 'periodic' cache strategy" ),
+      QStringLiteral( "Server" ), Qgis::MessageLevel::Info );
+  }
+  else if ( settings && settings->projectCacheStrategy() == QStringLiteral( "off" ) )
+  {
+    strategy = new QgsNullCacheStrategy();
+    QgsMessageLog::logMessage(
+      QStringLiteral( "Initializing 'off' cache strategy" ),
+      QStringLiteral( "Server" ), Qgis::MessageLevel::Info );
+  }
+  else
+  {
+    strategy = new QgsFileSystemCacheStrategy();
+    QgsMessageLog::logMessage(
+      QStringLiteral( "Initializing 'filesystem' cache strategy" ),
+      QStringLiteral( "Server" ), Qgis::MessageLevel::Info );
+  }
+
+  sInstance = new QgsConfigCache( strategy );
+}
+
+QgsConfigCache *QgsConfigCache::instance()
+{
   if ( !sInstance )
+  {
+    QgsMessageLog::logMessage(
+      QStringLiteral( "Project's cache not initialized, using 'filesystem' as default strategy" ),
+      QStringLiteral( "Server" ), Qgis::MessageLevel::Info );
     sInstance = new QgsConfigCache();
```

Comment:
```suggestion
    Q_ASSERT( false )
    return nullptr;
```

Reply:
This should crash immediately. Otherwise CI will still be green and we'll happily merge and everyone on the planet lives with a couple of strange error messages in their logs.

Reply:
Not sure with that, IHMO this is an api break, I would like @elpaso opinion on this because afaik he is a consumer of the QgsConfigCache api.

Reply:
@elpaso are you ok if we foce require explicit initialization of the `QgsConfigCache` object prior to being able to instantiate it?
This is a direct consequence of the decision to make this a singleton.

Reply:
@dmarteau we are calling `QgsConfigCache.instance().removeEntry(path)` but only after a `QgsServer()` has been created so I guess there should not be any problem, can you confirm?

Reply:
`Init()` is called at server creation so it is ok in the server context. 

But my concern is about the fact that 'QgsConfigCache::instance()` is public api and can be legitimately used outside of the context of `QgsServer` instanciation in any python script. So, requiring that `QgsConfig::initialize(...)` must be called before any call to 'QgsConfigCache::instance()` will break these scripts.  So, in the absolute, this is an api break.

Reply:
Technically you are correct.
However, is there a real world use case where the first call to `instance()` is done from outside qgis core? If there is actually such a scenario, and some qgis server plugin does that and subsequently the qgis server core implementation calls `init()` because it has no way of knowing, what would be the desired behavior? I.e. technically speaking isn't that already an API break already?

FWIW: I think it's easy to get away with this if we document this as an API break, document it properly, make sure some warning flags are raised if someone unintentionally runs into this (i.e. crash right away with a meaningful message). On the other hand I am not very much in favor of trying to get some semi-stable backwards compatibility which will likely never even be used in the wild.

And since we are on this topic, this is the exact reason why I am being so picky with limiting public API to only what is actually useful and stable enough.

Reply:
> Technically you are correct. However, is there a real world use case where the first call to `instance()` is done from outside qgis core? If there is actually such a scenario, and some qgis server plugin does that and subsequently the qgis server core implementation calls `init()` because it has no way of knowing, what would be the desired behavior? I.e. technically speaking isn't that already an API break already?
> 

Speaking of plugins they are all called during request processing and QgsServer is guaranteed to be initialized.

Reply:
Is it safe to ignore any use of this class outside plugins? As I understand, it's intended to be used within plugins, and if someone wants to use it outside it's on his own risk?

Reply:
I can't see a use case for that, in any event it is enough to use it after QgsServer is initialized, crashing with a meaningful message is probably a good idea if that condition isn't met.

Reply:
It might be self-evident, so please be gentle with me if I'm just stating the obvious.

Crashing is actually a *good thing* because it leaves us with a stack trace at the exact position where something unexpected happens. Just like throwing an exception in python. If we "only" print an error, we may know that something has gone wrong if we actually do have a look at the log and even if we see it, we have no idea *how* we got to this point.

An alternative (in general) would be throwing an exception. In this case here I don't see where and for what reason we should catch that. And it would add considerable complexity (especially if we would want that to traverse to Python).

Reply:
@m-kuhn I think we are on the same page: something like this should work https://github.com/qgis/QGIS/blob/master/src/server/qgsserver.cpp#L74